### PR TITLE
Allow setting application name from config

### PR DESCRIPTION
**Pull request by johngmyers**
Allow setting application name from config

### CODE REVIEW

File: bootstrap/src/main/java/com/proofpoint/bootstrap/Bootstrap.java [hunk 1]
```diff
@@ -77,14 +78,18 @@
...
+        return new StaticBootstrapBeforeModules(applicationName);
     }
 
-    private Bootstrap(String applicationName, Iterable<? extends Module> modules, boolean initializeLogging)
+    public static <T> BootstrapBeforeModules bootstrapApplication(Class<T> configClass, Function<T, String> applicationNameFunction)
+    {
+        return new DynamicBootstrapBeforeModules<>(configClass, applicationNameFunction);
+    }
+
+    private Bootstrap(Module applicationNameModule, Iterable<? extends Module> modules, boolean initializeLogging)
```
Comment (gwittel) [inline bootstrap/src/main/java/com/proofpoint/bootstrap/Bootstrap.java]: This seems to just trust that the ```applicationNameModule``` does the right thing and configures an instance of the ```ApplicationNameModule``` somewhere.

Would we prefer to enforce this programmatically vs just letting the service fail during ```NodeInfo``` creation?

Comment (johngmyers) [inline bootstrap/src/main/java/com/proofpoint/bootstrap/Bootstrap.java]: That is a private constructor and all extant static factory methods are tested to create a NodeInfo with the right application name.

Comment (johngmyers) [inline bootstrap/src/main/java/com/proofpoint/bootstrap/Bootstrap.java]: This commit is obsolete.

Comment (gwittel) [inline bootstrap/src/main/java/com/proofpoint/bootstrap/Bootstrap.java]: Agreed; it still trusts whoever touches this constructor does the right thing including tests.  I'm not a fan of this, but given it is internal, I don't feel strongly enough about this to object further.

**Comment by gwittel** [comment]
LGTM

**Comment by johngmyers** [comment]
Rebased, added CHANGES, and merged

**Pull request closed without merging**